### PR TITLE
[beta] backport fix for #94502

### LINKED ISSUE #94502: regression: ICE -- bad ConstKind after monomorphizing
[Crater](https://github.com/rust-lang/rust/issues/94345) picked up this issue in a number of crates:

May be a duplicate of https://github.com/rust-lang/rust/issues/93002, but filing this as a dedicated bug with a rough list of affected crates in Crater.

* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/mpancia.aoc2020/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/gytis-ivaskevicius.ion-shell/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/redox-os.ion/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/AestheticIntegration.batsmt/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/BenF-B.Titan/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/DataTriny.Fortuna/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/Furetur.jex_vm/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/Kilerd.leetcode-rust/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/SSebigo.iris-classification/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/SolarLiner.minigc.rs/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/TheArtistGuy.advent_of_code_2021/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/ajfrantz.advent2018/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/alexmailo.stack/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/andersmartini.workis/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/arnasjozonis.lz78/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/b4skyx.logchain/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/bantic.cryptopals/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/baoziv587.LoxLang/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/bitonic.puzzlescript-rs/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/byronwasti.advent-of-code-2019/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/cameron-williams.zkru/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/carribus.rust_commander/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/catunlock.numbers-rust/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/christian-blades-cb.aoc-2015/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/compenguy.rtf2text/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/curiousbirds.mint/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/davidsullins.AdventOfRust2018/log.txt 
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/dcheatha.graph-sketchpad/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/deaz.adventofcode-2018/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/garnish-lang.garnish-core/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/guilgaly.advent-of-code-2020/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/hsmtkk.super-duper-octo-potato/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/janglada.yahtzee_upper/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/jcrowgey.aoc_2018/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/jfrikker.minebot/log.txt
* (and more, not listing all of them here; we will rerun crater to confirm all are fixed.)

ty::ConstKind::Error constructed but no error reported (we should check, may be a distinct bug; maybe https://github.com/rust-lang/rust/issues/94506):

* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/MatsDK.CLI_File_Explorer/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/Qvist30.AocRust/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.60-1/beta-2022-02-22/gh/Rickgg.me_myself/log.txt

```
[INFO] [stdout] error: internal compiler error: /rustc/0a4f984a87c7ba6c74ec3e78442fec955a419e32/compiler/rustc_codegen_ssa/src/mir/constant.rs:42:20: encountered bad ConstKind after monomorphizing: Error(DelaySpanBugEmitted(()))
[INFO] [stdout]   --> parse/src/day02.rs:21:73
[INFO] [stdout]    |
[INFO] [stdout] 21 |     let char_count = count_dict.get(&password_rule.character).unwrap_or(&0);
[INFO] [stdout]    |                                                                         ^^
[INFO] [stdout] 
```

### PR DESCRIPTION
**Pull request by lcnr**
this issue was fixed as part of #93368, so i extracted the change from there

closes #94502

### CODE REVIEW

File: src/test/ui/nll/lint-no-err.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// check-pass
+
+// mir borrowck previously incorrectly set `tainted_by_errors`
+// when buffering lints, which resulted in ICE later on,
+// see #94502.
```
[the remark this answers is not in the document]
Comment (lcnr) [reply src/test/ui/nll/lint-no-err.rs:5]: already planned to do that :laughing: #94553

**Comment by oli-obk** [comment]
Adding the betanominated label so that triage picks it up today

**Comment by Dylan-DPC** [comment]
failed on creating the [rollup](https://github.com/rust-lang/rust/pull/94632) 

@bors r-

**Comment by Dylan-DPC** [comment]
@bors r+ rollup=never 

(could have been due to the different branch)

**Comment by Mark-Simulacrum** [comment]
This is beta-targeted and should never be rolled up -- I've updated the title to reflect its target branch.

**Comment by lcnr** [comment]
I don't understand that error :sweat_smile:

**Comment by oli-obk** [comment]
> tidy error: /checkout/src/test/ui/nll/lint-no-err.rs: missing trailing newline

Not sure why the bot didn't extract that

**Comment by lcnr** [comment]
@bors r=oli-obk

aaah, that's what happens when manually copying file contents without rerunning the tests

**Pull request merged**